### DISCUSSION TITLE
Additional niceties from Ruby corelib/stdlib and even ActiveSupport-isms?

### DISCUSSION THREAD

**Discussion by jaredcwhite** [ruby2js/ruby2js#71] [Ideas]
I'm thinking of starting a list of some obvious niceties we could look at implementing from Ruby's corelib/stdlib, and maybe even popular libraries like ActiveSupport (.present? would be awesome, for example…I always hate how verbose simple string checking has to be in JS). I also think it's likely some will make sense as function filters directly and others might be better suited to implementing via additional NPM packages and imported in.

Maybe using the wiki here for a page to start with would make sense…~~or keep a running list in this issue.…~~
https://github.com/rubys/ruby2js/wiki/Ruby-methods-possibly-supported-in-the-future

**Comment by rubys** [comment]
Just thinking out loud here... .assuming one or more of these results in a non-trivial expansion, it might make sense to create npm modules (perhaps under the `@ruby2js` scope/namespace?).  A filter might not be needed, but perhaps it could do whatever transformation is needed (if any) and insert the necessary `import` statements.

**Comment by jaredcwhite** [comment]
Definitely like the `@ruby2js` namespace idea, and an automatic import sounds interesting as well!

**Comment by rubys** [comment]
An example of automatic imports, try going to https://whimsy.apache.org/ruby2js/node/esm and entering the following

```ruby
system "echo hi"
```

The resulting javascript will be:

```
import child_process from "child_process";
child_process.execSync("echo hi", {stdio: "inherit"})
```

**Comment by jaredcwhite** [comment]
Very cool!

**Comment by rubys** [comment]
As luck would have it, npm already has a SecureRandom.alphanumeric implementation in [base62-random](https://github.com/deployable/base62-random).  I've committed experimental support for this (without docs) in https://github.com/rubys/ruby2js/commit/6e11f3122e57dc85edd89ed3c937357a0b95329c

Disjointed thoughts:

* the code to add CJS `require`s and ESM `import`s isn't very DRY.  That likely should be refactored by having all code generate `s(:import)` nodes, and having the converter for import do the right thing; `cjs` and ``esm` could be an option, and the `node` and `esm` filters could influence the default for this option, probably with the default ending up being esm UNLESS the node filter is used AND the esm filter is NOT used.
* While this an similar filters can be individually included, there may be value in having an umbrella `stdlib` filter that combines all of relevant filters.
* There may be value in tracking the imports that are found and having the object that `Ruby2js.convert` returns have a method that returns this list.  
* while `base64_random` is relatively unlikely to be collision with variable names in the converted script, it isn't impossible for it to be a collision.  Over time, as more modules are created, this chance goes up.  Perhaps a convention like adding a trailing `$` to such variable names might be worth exploring, as such would never be a valid Ruby identifier.

**Comment by jaredcwhite** [comment]
I like all of those ideas, and I would *love* to have a more official subsystem for imports, not only on the tracking output side as you mentioned, but I could also envision setting up some autoimports. Not necessarily something full-blown ala Zeitwerk, but maybe a situation where the converter is passed a hash of classes/methods -> paths and then when it encounters that class or method used in a file it makes sure that gets imported from elsewhere in the source tree. For example I write a lot of files where I always have to import three things from one particular file, and if that was just assumed dynamically it would clean things up very nicely. It's the kind of stuff that gives an even more Ruby-esque feel.

**Comment by rubys** [comment]
> but maybe a situation where the converter is passed a hash of classes/methods -> paths and then when it encounters that class or method used in a file it makes sure that gets imported from elsewhere in the source tree

Something like https://github.com/rubys/ruby2js/commit/dacda5bc06d20178446a21294b634aa33dd41654 perhaps?

**Comment by jaredcwhite** [comment]
Just added a new `active_functions` filter along with a supporting NPM package (tiny, zero-dependency) where we can hang a few of the ActiveSupport-inspired functions. Starting out with `blank?`, `present?`, and `presence`

**Comment by jaredcwhite** [comment]
JS working example: https://codepen.io/jaredcwhite/pen/MWjvYgb

**Comment by delonnewman** [comment]
I've been doing some work on a DateRange TS/JS class on a private project, which got me thinking about implementing Ruby's Range protocol as a Range class with concrete subclasses for DateRange, NumberRange, StringRange etc. That could be used from JS but also have a Ruby2JS interface with a filter that could provide those features in a more Rubyish way.

Would that fit here? I've already started working on Date interface that would mirror Ruby.

I'm also thinking infinite ranges could be supported using generators where available.

**Reply by rubys** [reply]
Short answer: yes.  Definitely.

It probably should be a separate package from active_support, but it definitely fits in the `@ruby2js` npm namespace: https://www.npmjs.com/search?q=%40ruby2js

**Reply by delonnewman** [reply]
I've also got some JS code that could be retrofitted to go into the `active_functions` library / filter it parses url params and form fields name / values into a nested hash like Rails / Rack on the client side.

See
- [params](https://github.com/delonnewman/js-helpers/blob/128d2cd031a331f8d6884ebbb1aaa2cd864fec2e/vendor/assets/javascripts/js-helpers.js#L667)
- [parseParamStr](https://github.com/delonnewman/js-helpers/blob/master/vendor/assets/javascripts/js-helpers.js#L812)
- [parseParamKey](https://github.com/delonnewman/js-helpers/blob/master/vendor/assets/javascripts/js-helpers.js#L739)
- [formData](https://github.com/delonnewman/js-helpers/blob/128d2cd031a331f8d6884ebbb1aaa2cd864fec2e/vendor/assets/javascripts/js-helpers.js#L682)

There are a [few other functions](https://github.com/delonnewman/js-helpers/blob/128d2cd031a331f8d6884ebbb1aaa2cd864fec2e/vendor/assets/javascripts/js-helpers.js#L667) that may be of interest as well like `timeAgoWithWords`, etc.

You can find them all here:  [delonnewman/js-helpers](https://github.com/delonnewman/js-helpers). It was a small library bundled as gem to provide a small stdlib that would (mostly) complement Rails and underscore.js that I've used on private projects.

**Reply by delonnewman** [reply]
> Short answer: yes. Definitely.
> 
> It probably should be a separate package from active_support, but it definitely fits in the `@ruby2js` npm namespace: https://www.npmjs.com/search?q=%40ruby2js

Yes, that makes sense. So once I'm ready would I just publish it on npm using `@ruby2js`? Or should I do a pull request on a particular repo?

I guess I'm not well educated on how npm namespaces work 😅.

**Reply by rubys** [reply]
Put the npm package here: https://github.com/ruby2js/ruby2js/tree/master/packages, and then we will work through giving you the permissions at that time.

**Reply by delonnewman** [reply]
Cool, that works. Thank you!

**Reply by jaredcwhite** [reply]
Sounds exciting @delonnewman.

**Reply by delonnewman** [reply]
Just so you don't feel like I've just fallen off the face of the earth. 😅 I'm still planning on working on this, in fact, I've given it a fair bit of thought, I've just been busy / dealing with some health issues.

**Reply by jaredcwhite** [reply]
didn't thumbs up the health issues (sorry to hear that), but the rest of it. =)
